### PR TITLE
Cockpit pcp python follow up

### PR DESCRIPTION
**Pull request by jelly**
Throwing this at CI, this still needs:

* Obsoletes/replaces of cockpit-pcp?
* Drop pkg/Makefile.am changes as they go in #20000

### CODE REVIEW

File: tools/debian/rules [hunk 1]
```diff
@@ -5,11 +5,7 @@ DEB_HOST_MULTIARCH ?= $(shell dpkg-architecture -qDEB_HOST_MULTIARCH)
 # we need an apparmor profile for >= 4.0; this exists in Ubuntu >= 24.04, but not in Debian yet
 PRE4AA = $(filter $(shell . /etc/os-release; echo $${VERSION_ID:-unstable}),22.04 11 12 unstable)
 
-# pcp isn't in Debian testing right now
-ifeq ($(shell . /etc/os-release; echo $${VERSION_CODENAME}),trixie)
-    export DH_OPTIONS = -Ncockpit-pcp
-    CONFIG_OPTIONS := --disable-pcp
-endif
+CONFIG_OPTIONS := --disable-pcp
```
Comment (martinpitt) [inline tools/debian/rules:8]: This is weird -- that option will go away entirely, and this PR should at least make that the default? Can you just leave it empty?

File: tools/debian/control [hunk 2]
```diff
@@ -92,15 +89,6 @@ Description: Cockpit user interface for networking
  The Cockpit components for interacting with networking configuration.
  Incompatible with systemd-networkd/ifupdown defined networks.
 
-Package: cockpit-pcp
```
Comment (martinpitt) [inline tools/debian/control:95]: Please add `Conflicts:/Replaces/Provides: cockpit-pcp` to cockpit-bridge.

https://www.debian.org/doc/debian-policy/ch-relationships.html#replacing-whole-packages-forcing-their-removal

File: tools/cockpit.spec [hunk 3]
```diff
@@ -588,24 +565,6 @@ These files are not required for running Cockpit.
 %{_unitdir}/cockpit-session.socket
 %{_unitdir}/cockpit-session@.service
 
-%if %{build_pcp}
-
-%package -n cockpit-pcp
```
Comment (martinpitt) [inline tools/cockpit.spec:593]: Please add `Obsoletes: cockpit-pcp < 326` to cockpit-bridge.

https://docs.fedoraproject.org/en-US/packaging-guidelines/#renaming-or-replacing-existing-packages

File: pkg/metrics/metrics.jsx [hunk 4]
```diff
@@ -1381,8 +1386,11 @@ const PCPConfigDialog = ({
     const handleInstall = async () => {
     // when enabling services, install missing packages on demand
         const missing = [];
-        if (dialogLoggerValue && !s_pmlogger.exists)
-            missing.push("cockpit-pcp");
+        let pcp_missing = false;
+        if (dialogLoggerValue && !s_pmlogger.exists) {
+            missing.push(await get_pcp_packagelist());
```
Comment (martinpitt) [inline pkg/metrics/metrics.jsx:1391]: So this assumes that pcp always depends on python3-pcp, right? i.e `!s_pmlogger.exists` is a sufficient condition.

File: pkg/metrics/metrics.jsx [hunk 5]
```diff
@@ -238,6 +238,11 @@ function make_rows(rows, rowProps, columnLabels) {
     );
 }
 
+async function get_pcp_packagelist() {
+    const os_release = await read_os_release();
+    return get_manifest_config_matchlist("metrics", "python_pcp", "python3-pcp", [os_release.PLATFORM_ID, os_release.ID]);
```
Comment (martinpitt) [inline pkg/metrics/metrics.jsx:243]: So while I get that `pcp` depends on `python3-pcp`, the opposite is probably not true? I.e. installing python3-pcp doesn't guarantee that `pmlogger.service` will come into existence?

File: test/verify/check-client [hunk 6]
```diff
@@ -37,14 +37,20 @@ class TestClient(testlib.MachineCase):
         self.m_target.execute("hostnamectl set-hostname target")
         # validate on-demand install: this does not work on arch, non-split package
         if self.m_target.image.startswith("debian") or self.m_target.image.startswith("ubuntu"):
-            self.m_target.execute("dpkg --purge cockpit-pcp-dbgsym || true; dpkg --purge cockpit-pcp pcp")
+            self.m_target.execute("""
+                if dpkg -l python3-pcp; then
```
Comment (martinpitt) [inline test/verify/check-client:41]: When would that ever not be true? If pcp depends on python3-pcp, and the else branch removes pcp, then python3-pcp should always be installed, too.

**Review by martinpitt — CHANGES_REQUESTED** [review]
Thanks! I'm really excited about this, it'll solve some long-standing downstream packaging trouble.

File: pkg/metrics/metrics.jsx (hunk 5 above)
Comment (jelly) [reply pkg/metrics/metrics.jsx:243]: I thought python3-pcp => pcp-libs => pcp is how the dependency chain worked. Let's check.

File: test/verify/check-client (hunk 6 above)
Comment (jelly) [reply test/verify/check-client:41]: It's a fallback as we afaik have no python3-pcp on debian-testing:

```
root@debian-testing-127-0-0-2-2201:~# dpkg -l | grep python3-pcp
root@debian-testing-127-0-0-2-2201:~#
```

Comment (jelly) [reply test/verify/check-client:41]: But yeah it should also remove pcp...

File: tools/debian/rules (hunk 1 above)
Comment (jelly) [reply tools/debian/rules:8]: WDYM? It doesn't seem to make sense to make disable-pcp enable-pcp in this PR? I thought we discussed that, if that needs to happen I can ofcourse make that happen.

Comment (martinpitt) [reply tools/debian/rules:8]: Not sure what you mean with "make disable-pcp enable-pcp". Oh, is that because we still have all the autoconfery and the source code in main? Then it's fine of course.

File: pkg/metrics/metrics.jsx (hunk 5 above)
Comment (martinpitt) [reply pkg/metrics/metrics.jsx:243]: E.g. in Debian, [pcp](https://packages.debian.org/sid/pcp) depends on [python3-pcp](https://packages.debian.org/sid/python3-pcp), but not the other way around. A mutual dependency is broken. Conceptually, the app (pcp) depends on the lib, but conceptually a Python module to interface with pcp data (e.g. archive files) mustn't pull in systemd services and such.

As `cockpit-pcp` depends on `pcp`, I think it'd make sense to keep that behaviour for now -- i.e. install `pcp` on demand, not `python3-pcp`. That even gets rid of the arch special-case.

File: tools/arch/PKGBUILD [hunk 7]
```diff
@@ -36,13 +36,16 @@ build() {
...
   make all
 }
 
 package_cockpit() {
   depends=(krb5 libssh json-glib glib-networking python)
+  provides=(cockpit-pcp)
+  replaces=(cockpit-pcp)
   backup=('etc/pam.d/cockpit' 'etc/cockpit/disallowed-users')
-  optdepends=("cockpit-pcp: reading performance metrics"
+  optdepends=("python3-pcp: reading performance metrics"
```
Comment (martinpitt) [inline tools/arch/PKGBUILD:48]: Feels like this should also be `pcp`, to get the full app?

File: test/verify/check-client (hunk 6 above)
Comment (martinpitt) [reply test/verify/check-client:41]: IMHO it would make things a lot simpler to just deal with the `pcp` package and nothing else. Here in the test, in the packaging dependencies, and in the on-demand install. The latter already tests for the `.service` (which is pcp, not python3-pcp). That would make everything consistent and simpler.

File: tools/debian/rules (hunk 1 above)
Comment (jelly) [reply tools/debian/rules:8]: It is, PCP is enabled by default.

File: pkg/metrics/metrics.jsx [hunk 8]
```diff
@@ -1381,8 +1381,11 @@ const PCPConfigDialog = ({
     const handleInstall = async () => {
     // when enabling services, install missing packages on demand
         const missing = [];
-        if (dialogLoggerValue && !s_pmlogger.exists)
-            missing.push("cockpit-pcp");
+        let pcp_missing = false;
+        if (dialogLoggerValue && !s_pmlogger.exists) {
+            missing.push("pcp");
+            pcp_missing = true;
```
Comment (cockpituous) [inline pkg/metrics/metrics.jsx:1387]: These 3 added lines are not executed by any test.

File: pkg/metrics/metrics.jsx [hunk 9]
```diff
@@ -1395,7 +1398,7 @@ const PCPConfigDialog = ({
             Dialogs.close();
             await install_dialog(missing);
             debug("PCPConfig: package installation successful");
-            if (missing.indexOf("cockpit-pcp") >= 0)
+            if (pcp_missing)
```
Comment (cockpituous) [inline pkg/metrics/metrics.jsx:1401]: This added line is not executed by any test.

File: pkg/metrics/metrics.jsx [hunk 10]
```diff
@@ -1824,14 +1842,11 @@ class MetricsHistory extends React.Component {
...
+        if (!this.state.isPythonPCPInstalled || this.pmlogger_service.exists === false)
             return <EmptyStatePanel
                         icon={ExclamationCircleIcon}
-                        title={_("Package cockpit-pcp is missing for metrics history")}
-                        action={this.state.isBeibootBridge === true
-                            // See https://github.com/cockpit-project/cockpit/issues/19143
-                            ? <Text>{ _("Installation not supported without installed cockpit package") }</Text>
-                            : this.state.packagekitExists && <Button onClick={this.handleInstall}>{_("Install cockpit-pcp")}</Button>}
+                        title={_("Python PCP module is missing for metrics history")}
+                        action={this.state.packagekitExists && <Button onClick={this.handleInstall}>{_("Install Python PCP module")}</Button>}
```
Comment (martinpitt) [inline pkg/metrics/metrics.jsx:1849]: same

File: src/cockpit/channels/pcp.py [hunk 11]
```diff
@@ -660,6 +660,10 @@ async def run(self, options: JsonObject) -> None:
 
         self.parse_options(options)
         try_import_pcp()
+        # HACK: the mock package test only sets pmapi to true
+        if pmapi is True:  # pragma: no cover
+            raise ChannelError('not-supported')
```
Comment (martinpitt) [inline src/cockpit/channels/pcp.py:665]: I don't understand this -- the "mock" is the little rpm that the integration test creates for on-demand install? If that sets `pmapi`, then shouldn't that count as "supported" (as the package exists), not "not-supported"?

File: test/verify/check-client [hunk 12]
```diff
@@ -37,14 +37,14 @@ class TestClient(testlib.MachineCase):
         self.m_target.execute("hostnamectl set-hostname target")
         # validate on-demand install: this does not work on arch, non-split package
         if self.m_target.image.startswith("debian") or self.m_target.image.startswith("ubuntu"):
-            self.m_target.execute("dpkg --purge cockpit-pcp-dbgsym || true; dpkg --purge cockpit-pcp pcp")
+            self.m_target.execute("dpkg --purge pcp")
         elif self.m_target.image != 'arch':
             # TODO: remove conditional when all images have python3-pcp and a Python PCP bridge
```
Comment (martinpitt) [inline test/verify/check-client:42]: Are there any Fedora/RHEL images which don't?

File: test/verify/check-client [hunk 13]
```diff
@@ -37,14 +37,14 @@ class TestClient(testlib.MachineCase):
...
             self.m_target.execute("""
                 if rpm -q python3-pcp; then
-                    rpm --erase --verbose cockpit-pcp pcp python3-pcp
+                    rpm --erase --verbose pcp python3-pcp
```
Comment (martinpitt) [inline test/verify/check-client:45]: Why is this necessary? AFAICS, the code checks for the .service. I suppose the bridge channel checks for the module, and that somehow interferes?

Or is that because `rpm --erase` would fail with just "pcp" (i.e. you need to explicitly remove the dependency)? in that case I don't understand the "else" branch.

File: pkg/metrics/metrics.jsx [hunk 14]
```diff
@@ -238,6 +238,18 @@ function make_rows(rows, rowProps, columnLabels) {
...
 }
 
+async function get_pcp_packages() {
+    const os_release = await read_os_release();
+    const pcp_packages = ["pcp"];
+
+    // PCP contains the Python module on Arch Linux, for all other distro's it is split up.
+    if (os_release.ID !== "arch") {
+        pcp_packages.push("python3-pcp");
+    }
```
[the remark this answers is not in the document]
Comment (jelly) [reply pkg/metrics/metrics.jsx:248]: But on Fedora, PCP simply does not depend on python3-pcp?

```
[root@fedora-40-127-0-0-2-2201 ~]# sudo dnf install -y pcp
Updating Subscription Management repositories.
Unable to read consumer identity

This system is not registered with an entitlement server. You can use subscription-manager to register.

Fedora 40 - x86_64                                                                                                                                              4.4 MB/s |  20 MB     00:04
Fedora 40 openh264 (From Cisco) - x86_64                                                                                                                        1.0 kB/s | 1.4 kB     00:01
Fedora 40 - x86_64 - Updates                                                                                                                                    6.6 MB/s |  10 MB     00:01
Dependencies resolved.
================================================================================================================================================================================================
 Package                                         Architecture                               Version                                           Repository                                   Size
================================================================================================================================================================================================
Installing:
 pcp                                             x86_64                                     6.3.1-1.fc40                                      updates                                     1.5 M
Installing dependencies:
 pcp-conf                                        x86_64                                     6.3.1-1.fc40                                      updates                                      30 k
 pcp-libs                                        x86_64                                     6.3.1-1.fc40                                      updates                                     650 k
 pcp-selinux                                     x86_64                                     6.3.1-1.fc40                                      updates                                      31 k

Transaction Summary
================================================================================================================================================================================================
Install  4 Packages

Total download size: 2.2 M
Installed size: 6.8 M
Downloading Packages:
(1/4): pcp-conf-6.3.1-1.fc40.x86_64.rpm                                                                                                                          60 kB/s |  30 kB     00:00
(2/4): pcp-libs-6.3.1-1.fc40.x86_64.rpm                                                                                                                         368 kB/s | 650 kB     00:01
(3/4): pcp-selinux-6.3.1-1.fc40.x86_64.rpm                                                                                                                       16 kB/s |  31 kB     00:02
(4/4): pcp-6.3.1-1.fc40.x86_64.rpm                                                                                                                              563 kB/s | 1.5 MB     00:02
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Total                                                                                                                                                           648 kB/s | 2.2 MB     00:03
Running transaction check
Transaction check succeeded.
Running transaction test
Transaction test succeeded.
Running transaction
  Preparing        :                                                                                                                                                                        1/1
  Running scriptlet: pcp-selinux-6.3.1-1.fc40.x86_64                                                                                                                                        1/4
  Installing       : pcp-selinux-6.3.1-1.fc40.x86_64                                                                                                                                        1/4
  Running scriptlet: pcp-selinux-6.3.1-1.fc40.x86_64                                                                                                                                        1/4
libsemanage.semanage_direct_install_info: Overriding pcp module at lower priority 100 with module at priority 200.

  Installing       : pcp-conf-6.3.1-1.fc40.x86_64                                                                                                                                           2/4
  Installing       : pcp-libs-6.3.1-1.fc40.x86_64                                                                                                                                           3/4
  Running scriptlet: pcp-6.3.1-1.fc40.x86_64                                                                                                                                                4/4
  Installing       : pcp-6.3.1-1.fc40.x86_64                                                                                                                                                4/4
  Running scriptlet: pcp-6.3.1-1.fc40.x86_64                                                                                                                                                4/4
Created symlink /etc/systemd/system/multi-user.target.wants/pmcd.service → /usr/lib/systemd/system/pmcd.service.
Created symlink /etc/systemd/system/multi-user.target.wants/pmlogger.service → /usr/lib/systemd/system/pmlogger.service.
Created symlink /etc/systemd/system/multi-user.target.wants/pmie.service → /usr/lib/systemd/system/pmie.service.

Installed products updated.

Installed:
  pcp-6.3.1-1.fc40.x86_64                    pcp-conf-6.3.1-1.fc40.x86_64                    pcp-libs-6.3.1-1.fc40.x86_64                    pcp-selinux-6.3.1-1.fc40.x86_64

Complete!
[root@fedora-40-127-0-0-2-2201 ~]# python3 -c 'import pcp'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'pcp'
```


But on Debian it does.

File: pkg/metrics/metrics.jsx [hunk 15]
```diff
@@ -1595,6 +1610,7 @@ class MetricsHistory extends React.Component {
             selectedDate: null,
             packagekitExists: false,
             isBeibootBridge: false,
+            isPythonPCPInstalled: false,
```
[the remark this answers is not in the document]
Comment (jelly) [reply pkg/metrics/metrics.jsx:1613]: Uh, yes?

File: src/cockpit/channels/pcp.py (hunk 11 above)
Comment (jelly) [reply src/cockpit/channels/pcp.py:665]: So our test installed a fake `cockpit-pcp` package which contained a valid manifest which made the page no longer show "package missing" but instead shows "no metrics".

As our mock package installs a pcp module to show handle "no package" situation, it would then crash and require an `allow_journal_message` for the exception. Instead of allowing that exception, I chose this solution

File: test/verify/check-client (hunk 12 above)
Comment (martinpitt) [reply test/verify/check-client:42]: Sure, follow-up is fine, as long as you keep a list :wink:

File: pkg/metrics/metrics.jsx [hunk 16]
```diff
@@ -238,6 +238,18 @@ function make_rows(rows, rowProps, columnLabels) {
     );
 }
 
+async function get_pcp_packages() {
```
Comment (martinpitt) [inline pkg/metrics/metrics.jsx:241]: Commit message: Please say "pcp" (or both), as it's really not just about the Python binding.

File: pkg/metrics/metrics.jsx [hunk 17]
```diff
@@ -238,6 +238,18 @@ function make_rows(rows, rowProps, columnLabels) {
...
+    const os_release = await read_os_release();
+    const pcp_packages = ["pcp"];
+
+    // PCP contains the Python module on Arch Linux, for all other distro's it is split up.
```
Comment (martinpitt) [inline pkg/metrics/metrics.jsx:245]: For our future selves: Please document why we have to explicitly install both: That on Fedora, python3-pcp depends on pcp, and on Debian pcp depends on python3-pcp.

File: pkg/metrics/metrics.jsx (hunk 15 above)
Comment (martinpitt) [inline pkg/metrics/metrics.jsx:1613]: This smells like generating flicker. Let's initialize to `null`, and fix the comparison below.

File: pkg/metrics/metrics.jsx [hunk 18]
```diff
@@ -1789,8 +1805,10 @@ class MetricsHistory extends React.Component {
                 this.setState({
                     loading: false,
                     metricsAvailable: false,
+                    isPythonPCPInstalled: message?.message !== "python3-pcp not installed",
```
Comment (martinpitt) [inline pkg/metrics/metrics.jsx:1808]: Please do `message.problem === "not-supported"` ? That's much better than this human string comparison. Also, your mock hack just does `raise ChannelError('not-supported')` without a message.

(The `?` here is unnecessary -- you already asserted that message.problem exists)

File: pkg/metrics/metrics.jsx (hunk 10 above)
Comment (martinpitt) [inline pkg/metrics/metrics.jsx:1849]: Let's please not pick on "Python" - just say "PCP is missing". This installs both, and the main thing is really the `pcp` package/service.

File: pkg/metrics/metrics.jsx [hunk 19]
```diff
@@ -1824,14 +1842,11 @@ class MetricsHistory extends React.Component {
 
         // on a single machine, cockpit-pcp depends on pcp; but this may not be the case in the beiboot scenario,
         // so additionally check if pcp is available on the logged in target machine
-        if ((cockpit.manifests && !cockpit.manifests.pcp) || this.pmlogger_service.exists === false)
+        if (!this.state.isPythonPCPInstalled || this.pmlogger_service.exists === false)
```
Comment (martinpitt) [inline pkg/metrics/metrics.jsx:1845]: See above, please compare to `=== false` to not show this message while initializing the page.

File: test/verify/check-client [hunk 20]
```diff
@@ -143,21 +143,16 @@ exec "$@"
         self.check_login("admin@target")
         b.become_superuser()
 
-        # cockpit-pcp is installed on client, but not on target; recognize that
+        # python3-pcp is installed on client, but not on target; recognize that
         if self.m_target.image != 'arch':
```
Comment (martinpitt) [inline test/verify/check-client:147]: This should now also work on arch, right? OK to be a follow-up, but this PR changes arch code, so it needs to be tested at least manually, or the follow-up is a release blocker.

**Review by martinpitt — CHANGES_REQUESTED** [review]
I re-reviewed this from scratch after the previous discussions about the weird dependency direction in Fedora.

File: test/verify/check-client (hunk 20 above)
Comment (jelly) [reply test/verify/check-client:147]: It should work fine, I'm happy to do this in a follow up

File: pkg/metrics/metrics.jsx (hunk 18 above)
Comment (jelly) [reply pkg/metrics/metrics.jsx:1808]: No, the mock is for a different scenario.

File: test/verify/check-client (hunk 20 above)
Comment (jelly) [reply test/verify/check-client]: Sneaked this in as last commit.

**Review by martinpitt — APPROVED** [review]
I really dislike the human error string comparison instead of the problem code, but I'm ok with that being a follow-up. Thanks!
